ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,276 @@
+             in
+             let versions_changed =
+               Db.Host_driver.get_versions ~__context ~self <> driver_versions
+             in
+             if selected_changed then
+               Db.Host_driver.set_selected_version ~__context ~self
+                 ~value:selected_version ;
+             if active_changed then
+               Db.Host_driver.set_active_version ~__context ~self
+                 ~value:active_version ;
+             if versions_changed then
+               Db.Host_driver.set_versions ~__context ~self
+                 ~value:driver_versions ;
+             if selected_changed || active_changed || versions_changed then
+               Some driver_name
+             else
+               None
+         )
+    in
+    (* Remove drivers that are no longer in the filesystem from the DB *)
+    let removed_drivers =
+      DriverSet.diff db_drivers_set filesystem_drivers
+      |> DriverSet.elements
+      |> List.map (fun driver_name ->
+             let driver = DriverMap.find driver_name db_drivers_map in
+             destroy ~__context ~self:driver.ref ;
+             driver_name
+         )
+    in
+    (* Add the new drivers *)
+    let added_drivers =
+      DriverSet.diff filesystem_drivers db_drivers_set
+      |> DriverSet.elements
+      |> List.map (fun driver_name ->
+             let selected_version, active_version, driver_versions =
+               __get_driver_versions driver_name driver_paths.selected_dir
+                 driver_paths.all_versions_dir
+             in
+             create ~__context ~host ~name_label:driver_name
+               ~versions:driver_versions ~active_version ~selected_version ;
+             driver_name
+         )
+    in
+    [
+      ("updated_drivers", updated_drivers)
+    ; ("removed_drivers", removed_drivers)
+    ; ("added_drivers", added_drivers)
+    ]
+    |> List.filter_map (fun (name, lst) ->
+           match lst with _ :: _ as lst -> Some (name, lst) | _ -> None
+       )
+  with
+  | Sys_error e ->
+      D.debug "Failed to parse drivers on host %s: %s" (Ref.string_of host) e ;
+      []
+  | e ->
+      D.debug "Failed to parse drivers on host %s: %s - %s" (Ref.string_of host)
+        (Printexc.to_string e)
+        (Printexc.get_backtrace ()) ;
+      []
+
+(* Runs on the necessary host *)
+let select ~__context ~self ~version =
+  let versions = Db.Host_driver.get_versions ~__context ~self in
+  match List.find_opt (String.equal version) versions with
+  | Some _ -> (
+      let host = Helpers.get_localhost ~__context in
+      let driver_name = Db.Host_driver.get_name_label ~__context ~self in
+      let driver_paths = __get_drivers_base_path ~__context ~host in
+      let srcfile =
+        driver_paths.all_versions_dir
+        // driver_name
+        // version
+        // (driver_name ^ ".ko")
+      in
+      let destfile = driver_paths.selected_dir // (driver_name ^ ".ko") in
+      try
+        if Sys.file_exists destfile then
```

Comment:
It does not matter here, but this created a race between test and action. It's better to always remove and to catch the failure.